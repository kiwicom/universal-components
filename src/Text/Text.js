// @flow

import * as React from 'react';
import { Text as RNText, Platform } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import StyleSheet from '../PlatformStyleSheet';
import { createStylesGenerator } from '../utils';
import {
  align as alignTypes,
  fontSize,
  fontWeight as fontWeightTypes,
  textColor,
} from './styles';

import type { TextType } from './TextTypes';

const colorGen = createStylesGenerator('color', textColor);
const fontWeightGen = createStylesGenerator('fontWeight', fontWeightTypes);
const fontSizeGen = createStylesGenerator('fontSize', fontSize);
const alignGen = createStylesGenerator('textAlign', alignTypes);

const Text = ({
  align = 'left',
  children,
  dataTest,
  italic,
  numberOfLines,
  size = 'normal',
  style,
  type = 'primary',
  uppercase,
  weight = 'normal',
}: TextType) => (
  <RNText
    data-test={dataTest}
    style={[
      styles.text,
      italic && styles.italic,
      uppercase && styles.uppercase,
      styles[weight],
      styles[size],
      styles[type],
      styles[align],
      style && style,
    ]}
    numberOfLines={numberOfLines}
  >
    {children}
  </RNText>
);
const styles = StyleSheet.create({
  text: {
    margin: 0,
    fontFamily: Platform.OS === 'web' ? defaultTokens.fontFamily : 'Roboto',
  },
  italic: {
    fontStyle: 'italic',
  },
  uppercase: {
    textTransform: 'uppercase',
  },
  ...fontWeightGen(),
  ...alignGen(),
  ...colorGen(),
  ...fontSizeGen(),
});

export default Text;
