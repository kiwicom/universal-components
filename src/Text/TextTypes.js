// @flow

import * as React from 'react';

export type TextType = {|
  +align?: 'left' | 'right' | 'center' | 'justify',
  +children: React.Node,
  +dataTest?: string,
  +fontWeight?: 'normal' | 'bold',
  +italic?: boolean,
  +size?: 'normal' | 'large' | 'small',
  +uppercase?: boolean,
  +type?:
    | 'attention'
    | 'critical'
    | 'info'
    | 'primary'
    | 'secondary'
    | 'success'
    | 'warning'
    | 'white',
|};
