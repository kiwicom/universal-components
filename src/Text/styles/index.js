// @flow

import { Platform } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

export const shared = {
  margin: 0,
  fontFamily: Platform.OS === 'web' ? defaultTokens.fontFamily : 'Roboto',
};
