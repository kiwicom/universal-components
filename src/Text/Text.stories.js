// @flow

import React from 'react';
import { storiesOf } from '@storybook/react-native';
import Text from './index';

storiesOf('Text', module).add('default', () => (
  <Text>Godard keffiyeh Blue Bottle hoodie biodiesel leggings</Text>
));
