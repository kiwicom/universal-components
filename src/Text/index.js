// @flow

import * as React from 'react';
import { textColor, fontSize, fontWeight, align } from './styles';

const styles = {
  ...textColor,
  ...fontSize,
  ...fontWeight,
  ...align,
};

const Text = ({ element = 'p', children, type, ...props }) => {
  const TextElement = element;
  const style = { color: styles[type], fontFamily: 'Roboto' };
  return <TextElement style={style}>{children}</TextElement>;
};

export default Text;
