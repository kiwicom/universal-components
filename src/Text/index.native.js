// @flow

import * as React from 'react';
import { Text as RNText } from 'react-native';
import { StyleSheet } from '../index';
import { createStylesGenerator } from '../utils';
import {
  textAlign,
  fontSize,
  fontWeight as fontWeightTypes,
  textColor,
} from './constants';
import { shared } from './styles';

import type { TextType } from './TextTypes';

const colorGen = createStylesGenerator('color', textColor);
const fontWeightGen = createStylesGenerator('fontWeight', fontWeightTypes);
const fontSizeGen = createStylesGenerator('fontSize', fontSize);
const alingGen = createStylesGenerator('textAlign', textAlign);

const Text = ({
  children,
  italic,
  uppercase,
  dataTest,
  align = 'left',
  type = 'primary',
  weight = 'normal',
  size = 'normal',
}: TextType) => (
  <RNText
    data-test={dataTest}
    style={[
      styles.text,
      italic && styles.italic,
      uppercase && styles.uppercase,
      styles[weight],
      styles[size],
      styles[type],
      styles[align],
    ]}
  >
    {children}
  </RNText>
);

const styles = StyleSheet.create({
  text: {
    ...shared,
  },
  italic: {
    fontStyle: 'italic',
  },
  uppercase: {
    textTransform: 'uppercase',
  },
  ...fontWeightGen(),
  ...alingGen(),
  ...colorGen(),
  ...fontSizeGen(),
});

export default Text;
