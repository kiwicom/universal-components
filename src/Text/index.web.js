// @flow

import * as React from 'react';
import styled from 'styled-components';
import { textColor, fontSize, fontWeight, textAlign } from './constants';
import { shared } from './styles';
import type { TextType, Element } from './TextTypes.js';

type Props = TextType & {| +element: Element |};

const createElement = ({ element, children, ...props }: Props) =>
  React.createElement(element, props, children);

const StyledText = styled(createElement)(
  [],
  ({ type, size, weight, align, uppercase, italic }) => ({
    ...shared,
    color: textColor[type],
    fontSize: fontSize[size],
    fontWeight: fontWeight[weight],
    textAlign: textAlign[align],
    fontStyle: italic && 'italic',
    textTransform: uppercase && 'uppercase',
  })
);

const Text = ({
  align = 'left',
  type = 'primary',
  size = 'normal',
  weight = 'normal',
  element = 'p',
  uppercase = false,
  italic = false,
  dataTest,
  children,
}: Props) => (
  <StyledText
    type={type}
    size={size}
    weight={weight}
    align={align}
    element={element}
    uppercase={uppercase}
    italic={italic}
    data-test={dataTest}
  >
    {children}
  </StyledText>
);

export default Text;
