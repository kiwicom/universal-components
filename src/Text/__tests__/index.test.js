// @flow

import * as React from 'react';
import { render } from 'react-native-testing-library';

import Text from '../index';

describe('Text', () => {
  const children = 'Lorem ipsum';
  const { getByText } = render(<Text>{children}</Text>);
  it('should have defined children text', () => {
    expect(getByText(children)).toBeDefined();
  });
});
