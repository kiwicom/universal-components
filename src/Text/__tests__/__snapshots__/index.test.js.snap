// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should match snapshot diff 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,24 +3,28 @@
      Array [
        Object {
          \\"fontFamily\\": \\"\\\\\\"Roboto\\\\\\", -apple-system, \\\\\\".SFNSText-Regular\\\\\\", \\\\\\"San Francisco\\\\\\", \\\\\\"Segoe UI\\\\\\", \\\\\\"Helvetica Neue\\\\\\", \\\\\\"Lucida Grande\\\\\\", sans-serif\\",
          \\"margin\\": 0,
        },
-       undefined,
-       undefined,
        Object {
-         \\"fontSize\\": 14,
+         \\"fontStyle\\": \\"italic\\",
        },
        Object {
-         \\"fontSize\\": 14,
+         \\"textTransform\\": \\"uppercase\\",
        },
        Object {
-         \\"color\\": \\"#46515e\\",
+         \\"fontWeight\\": \\"700\\",
        },
        Object {
-         \\"textAlign\\": \\"left\\",
+         \\"fontSize\\": 16,
        },
+       Object {
+         \\"color\\": \\"#7f91a8\\",
+       },
+       Object {
+         \\"textAlign\\": \\"center\\",
+       },
      ]
    }
  >
    Lorem ipsum
  </Text>"
`;
